File: dbflow/src/main/java/com/raizlabs/android/dbflow/runtime/ContentResolverNotifier.java
package com.raizlabs.android.dbflow.runtime;

import android.content.ContentResolver;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.raizlabs.android.dbflow.config.FlowManager;
import com.raizlabs.android.dbflow.sql.SqlUtils;
import com.raizlabs.android.dbflow.sql.language.SQLOperator;
import com.raizlabs.android.dbflow.structure.BaseModel;
import com.raizlabs.android.dbflow.structure.ModelAdapter;

/**
 * The default use case, it notifies via the {@link ContentResolver} system.
 */
public class ContentResolverNotifier implements ModelNotifier {

    @NonNull
    private final String contentAuthority;

    public ContentResolverNotifier(@NonNull String contentAuthority) {
        this.contentAuthority = contentAuthority;
    }

    @Override
    public <T> void notifyModelChanged(@NonNull T model, @NonNull ModelAdapter<T> adapter,
                                       @NonNull BaseModel.Action action) {
        if (FlowContentObserver.shouldNotify()) {
            FlowManager.getContext().getContentResolver()
                    .notifyChange(SqlUtils.getNotificationUri(contentAuthority,
                            adapter.getModelClass(), action,
                            adapter.getPrimaryConditionClause(model).getConditions()), null, true);
        }
    }

    @Override
    public <T> void notifyTableChanged(@NonNull Class<T> table, @NonNull BaseModel.Action action) {
        if (FlowContentObserver.shouldNotify()) {
            FlowManager.getContext().getContentResolver()
                    .notifyChange(SqlUtils.getNotificationUri(contentAuthority,
                            table, action, (SQLOperator[]) null), null, true);
        }
    }

    @Override
    public TableNotifierRegister newRegister() {
        return new FlowContentTableNotifierRegister(contentAuthority);
    }

    public static class FlowContentTableNotifierRegister implements TableNotifierRegister {
        private final FlowContentObserver flowContentObserver;

        @Nullable
        private OnTableChangedListener tableChangedListener;

        public FlowContentTableNotifierRegister(@NonNull String contentAuthority) {
            flowContentObserver = new FlowContentObserver(contentAuthority);
            flowContentObserver.addOnTableChangedListener(internalContentChangeListener);
        }

        @Override
        public <T> void register(@NonNull Class<T> tClass) {
            flowContentObserver.registerForContentChanges(FlowManager.getContext(), tClass);
        }

        @Override
        public <T> void unregister(@NonNull Class<T> tClass) {
            flowContentObserver.unregisterForContentChanges(FlowManager.getContext());
        }

        @Override
        public void unregisterAll() {
            flowContentObserver.removeTableChangedListener(internalContentChangeListener);
            this.tableChangedListener = null;
        }

        @Override
        public void setListener(@Nullable OnTableChangedListener contentChangeListener) {
            this.tableChangedListener = contentChangeListener;
        }

        @Override
        public boolean isSubscribed() {
            return !flowContentObserver.isSubscribed();
        }

        private final OnTableChangedListener internalContentChangeListener
                = new OnTableChangedListener() {

            @Override
            public void onTableChanged(@Nullable Class<?> tableChanged, @NonNull BaseModel.Action action) {
                if (tableChangedListener != null) {
                    tableChangedListener.onTableChanged(tableChanged, action);
                }
            }
        };
    }
}

File: dbflow/src/main/java/com/raizlabs/android/dbflow/runtime/ContentResolverNotifier.kt
package com.raizlabs.android.dbflow.runtime

import android.content.ContentResolver

import com.raizlabs.android.dbflow.config.FlowManager
import com.raizlabs.android.dbflow.sql.getNotificationUri
import com.raizlabs.android.dbflow.sql.language.SQLOperator
import com.raizlabs.android.dbflow.structure.BaseModel
import com.raizlabs.android.dbflow.structure.ModelAdapter

/**
 * The default use case, it notifies via the [ContentResolver] system.
 */
class ContentResolverNotifier : ModelNotifier {

    override fun <T : Any> notifyModelChanged(model: T, adapter: ModelAdapter<T>,
                                              action: BaseModel.Action) {
        if (FlowContentObserver.shouldNotify()) {
            FlowManager.context.contentResolver
                    .notifyChange(getNotificationUri(adapter.modelClass, action,
                            adapter.getPrimaryConditionClause(model).conditions), null, true)
        }
    }

    override fun <T : Any> notifyTableChanged(table: Class<T>, action: BaseModel.Action) {
        if (FlowContentObserver.shouldNotify()) {
            FlowManager.context.contentResolver
                    .notifyChange(getNotificationUri(table, action,
                            null as Array<SQLOperator>?), null, true)
        }
    }

    override fun newRegister(): TableNotifierRegister = FlowContentTableNotifierRegister()

    class FlowContentTableNotifierRegister : TableNotifierRegister {

        private val flowContentObserver = FlowContentObserver()

        private var tableChangedListener: OnTableChangedListener? = null

        private val internalContentChangeListener = object : OnTableChangedListener {
            override fun onTableChanged(table: Class<*>?, action: BaseModel.Action) {
                tableChangedListener?.onTableChanged(table, action)
            }
        }

        init {
            flowContentObserver.addOnTableChangedListener(internalContentChangeListener)
        }

        override fun <T> register(tClass: Class<T>) {
            flowContentObserver.registerForContentChanges(FlowManager.context, tClass)
        }

        override fun <T> unregister(tClass: Class<T>) {
            flowContentObserver.unregisterForContentChanges(FlowManager.context)
        }

        override fun unregisterAll() {
            flowContentObserver.removeTableChangedListener(internalContentChangeListener)
            this.tableChangedListener = null
        }

        override fun setListener(listener: OnTableChangedListener?) {
            this.tableChangedListener = listener
        }

        override val isSubscribed: Boolean
            get() = !flowContentObserver.isSubscribed
    }
}
